File: common/src/main/java/de/hype/bbsentials/shared/objects/minions/MinionItem.java
package de.hype.bbsentials.shared.objects.minions;

import de.hype.bbsentials.shared.constants.Collections;
import de.hype.bbsentials.shared.constants.MinionResourceItem;

import java.util.Map;

public enum MinionItem {
    AUTO_SMELTER("Auto Smelter") {
        @Override
        public MinionResourceItem convertItem(MinionResourceItem item) {
            if (item instanceof Collections.Foraging) return Collections.Mining.Coal;
            return super.convertItem(item);
        }
    },
    COMPACTOR("Compactor"),
    SUPER_COMPACTOR_3000("Super Compactor 3000"),
    DWARFEN_SUPER_COMPACTOR("Dwarfen Super Compactor") {
        @Override
        public MinionResourceItem convertItem(MinionResourceItem item) {
            return AUTO_SMELTER.convertItem(item);
        }
    },
    DIAMOND_SPREADING("Diamond Spreading") {
        @Override
        public void modifyDrops(double sum, Map<MinionResourceItem, Double> generated) {
            generated.put(Collections.Mining.Diamond, generated.getOrDefault(Collections.Mining.Diamond, 0D) + (sum / 10));
        }
    },
    POTATO_SPREADING(false, "Potato Spreading") {
        @Override
        public void modifyDrops(double sum, Map<MinionResourceItem, Double> generated) {
            generated.put(Collections.Farming.Potato, generated.getOrDefault(Collections.Farming.Potato, 0D) + (sum / 20));
        }
    },
    MINION_EXPANDER("Minion Expander") {
        @Override
        public Integer getMinionSpeedAdditive(Minions minion) {
            return 5;
        }
    },
    ENCHANTED_EGG("Enchanted Egg"),
    FLINT_SHOVEL("Flint Shovel"),
    FLYCATCHER(false, "Flycatcher") {
        @Override
        public Integer getMinionSpeedAdditive(Minions minion) {
            return 20;
        }
    },
    KRAMPUS_HELMET(false, "Krampus Helmet") {
        @Override
        public void modifyDrops(double sum, Map<MinionResourceItem, Double> generated) {
            generated.put(MinionResourceItem.UnusedMinionItems.RED_GIFT, generated.getOrDefault(MinionResourceItem.UnusedMinionItems.RED_GIFT, 0D) + (sum * 0.000045));
        }
    },
    LESSER_SOULFLOW_ENGINE("Lesser Soulflow Engine") {
        @Override
        public Map<MinionResourceItem, Integer> itemsGenerated(Map<MinionResourceItem, Integer> dropsGenerated, Integer minionActions, Minions minion) {
            for (Map.Entry<MinionResourceItem, Integer> collectionsIntegerEntry : dropsGenerated.entrySet()) {
                dropsGenerated.put(collectionsIntegerEntry.getKey(), collectionsIntegerEntry.getValue() / 2);
            }
            return dropsGenerated;
        }
    },
    SOULFLOW_ENGINE("Soulflow Engine") {
        @Override
        public double getMultiplier(Minions minion) {
            return 0.5;
        }
    },
    CORRUPT_SOIL("Corrupt Soil") {
        @Override
        public Map<MinionResourceItem, Integer> itemsGenerated(Map<MinionResourceItem, Integer> dropsGenerated, Integer minionActions, Minions minion) {
            if (minion.spawnsMobs()) dropsGenerated.put(Collections.Mining.Sulphur, minionActions);
            return dropsGenerated;
        }
    },
    ENCHANTED_SHEARS("Enchanted Shears"),
    BERBERIES_FUEL_INJECTOR("Berberis Fuel Injector") {
        @Override
        public Integer getMinionSpeedAdditive(Minions minion) {
            if (minion.getType() == MinionType.FARMING) return 15;
            return 0;
        }
    },
    SLEEPY_HOLLOW(false, "Sleepy Hollow") {
        @Override
        public Map<MinionResourceItem, Integer> itemsGenerated(Map<MinionResourceItem, Integer> dropsGenerated, Integer minionActions, Minions minion) {
            //dropsGenerated.put(PURPLE_CANY,(int) dropsGenerated.values().stream().mapToInt(v->v).sum()*0.00015);
            return dropsGenerated;
        }
    };

    public String displayName;
    public boolean bingoObtainable;


    MinionItem(String name) {
        this(true, name);
    }

    MinionItem(boolean bingoObtainable, String name) {
        this.displayName = name;
        this.bingoObtainable = bingoObtainable;
    }

    public Integer getMinionSpeedAdditive(Minions minion) {
        return 0;
    }

    public Map<MinionResourceItem, Integer> itemsGenerated(Map<MinionResourceItem, Integer> dropsGenerated, Integer minionActions, Minions minion) {
        return dropsGenerated;
    }

    public Map<MinionResourceItem, Integer> applyCompacting(Map<MinionResourceItem, Integer> drops, Minions minion) {
        return drops;
    }


    public MinionResourceItem convertItem(MinionResourceItem item) {
        return item;
    }

    public double getMultiplier(Minions minion) {
        return 1;
    }

    public void modifyDrops(double sum, Map<MinionResourceItem, Double> generated) {
    }
}

File: common/src/main/kotlin/de/hype/bingonet/shared/objects/minions/MinionItem.kt
package de.hype.bingonet.shared.objects.minions

import de.hype.bingonet.shared.constants.Collections
import de.hype.bingonet.shared.constants.Collections.Foraging
import de.hype.bingonet.shared.constants.MinionResourceItem

enum class MinionItem(var bingoObtainable: Boolean, var displayName: String) {
    AUTO_SMELTER("Auto Smelter") {
        override fun convertItem(item: MinionResourceItem): MinionResourceItem {
            if (item is Foraging) return Collections.Mining.Coal
            return super.convertItem(item)
        }
    },
    COMPACTOR("Compactor"),
    SUPER_COMPACTOR_3000("Super Compactor 3000"),
    DWARFEN_SUPER_COMPACTOR("Dwarfen Super Compactor") {
        override fun convertItem(item: MinionResourceItem): MinionResourceItem {
            return AUTO_SMELTER.convertItem(item)
        }
    },
    DIAMOND_SPREADING("Diamond Spreading") {
        override fun modifyDrops(sum: Double, generated: MutableMap<MinionResourceItem, Double>) {
            generated.put(
                Collections.Mining.Diamond,
                generated.getOrDefault(Collections.Mining.Diamond, 0.0) + (sum / 10)
            )
        }
    },
    POTATO_SPREADING(false, "Potato Spreading") {
        override fun modifyDrops(sum: Double, generated: MutableMap<MinionResourceItem, Double>) {
            generated.put(
                Collections.Farming.Potato,
                generated.getOrDefault(Collections.Farming.Potato, 0.0) + (sum / 20)
            )
        }
    },
    MINION_EXPANDER("Minion Expander") {
        override fun getMinionSpeedAdditive(minion: Minions): Int {
            return 5
        }
    },
    ENCHANTED_EGG("Enchanted Egg"),
    FLINT_SHOVEL("Flint Shovel"),
    FLYCATCHER(false, "Flycatcher") {
        override fun getMinionSpeedAdditive(minion: Minions): Int {
            return 20
        }
    },
    KRAMPUS_HELMET(false, "Krampus Helmet") {
        override fun modifyDrops(sum: Double, generated: MutableMap<MinionResourceItem, Double>) {
            generated.put(
                MinionResourceItem.UnusedMinionItems.RED_GIFT,
                generated.getOrDefault(MinionResourceItem.UnusedMinionItems.RED_GIFT, 0.0) + (sum * 0.000045)
            )
        }
    },
    LESSER_SOULFLOW_ENGINE("Lesser Soulflow Engine") {
        override fun items(
            dropsGenerated: MutableMap<MinionResourceItem, Int>,
            minionActions: Int,
            minion: Minions
        ): MutableMap<MinionResourceItem, Int> {
            for (collectionsIntegerEntry in dropsGenerated.entries) {
                dropsGenerated.put(collectionsIntegerEntry.key, collectionsIntegerEntry.value / 2)
            }
            return dropsGenerated
        }
    },
    SOULFLOW_ENGINE("Soulflow Engine") {
        override fun getMultiplier(minion: Minions): Double {
            return 0.5
        }
    },
    CORRUPT_SOIL("Corrupt Soil") {
        override fun items(
            dropsGenerated: MutableMap<MinionResourceItem, Int>,
            minionActions: Int,
            minion: Minions
        ): MutableMap<MinionResourceItem, Int> {
            if (minion.spawnsMobs()) dropsGenerated.put(Collections.Mining.Sulphur, minionActions)
            return dropsGenerated
        }
    },
    ENCHANTED_SHEARS("Enchanted Shears"),
    BERBERIES_FUEL_INJECTOR("Berberis Fuel Injector") {
        override fun getMinionSpeedAdditive(minion: Minions): Int {
            if (minion.getType() == MinionType.FARMING) return 15
            return 0
        }
    },
    SLEEPY_HOLLOW(false, "Sleepy Hollow") {
        override fun items(
            dropsGenerated: MutableMap<MinionResourceItem, Int>,
            minionActions: Int,
            minion: Minions
        ): MutableMap<MinionResourceItem, Int> {
            //dropsGenerated.put(PURPLE_CANY,(int) dropsGenerated.values().mapToInt(v->v).sum()*0.00015);
            return dropsGenerated
        }
    };


    constructor(name: String) : this(true, name)

    open fun getMinionSpeedAdditive(minion: Minions): Int {
        return 0
    }

    open fun items(
        dropsGenerated: MutableMap<MinionResourceItem, Int>,
        minionActions: Int,
        minion: Minions
    ): MutableMap<MinionResourceItem, Int> {
        return dropsGenerated
    }

    fun applyCompacting(
        drops: MutableMap<MinionResourceItem, Int>,
        minion: Minions
    ): MutableMap<MinionResourceItem, Int> {
        return drops
    }


    open fun convertItem(item: MinionResourceItem): MinionResourceItem {
        return item
    }

    open fun getMultiplier(minion: Minions): Double {
        return 1.0
    }

    open fun modifyDrops(sum: Double, generated: MutableMap<MinionResourceItem, Double>) {}
}
